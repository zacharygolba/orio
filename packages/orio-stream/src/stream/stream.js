// @flow

import { AsAsyncIterator, ToString } from 'orio-traits'
import { identity, reduceAsync } from 'orio-utils'
import type {
  AsyncDest,
  AsyncProducer,
  AsyncIteratorResult,
  FromIterator,
  AsyncSource,
} from 'orio-types'

import * as sink from '../sink'
import type { Sink } from '../sink'

import {
  Chain,
  Concat,
  Cycle,
  Enumerate,
  Filter,
  FilterMap,
  FlatMap,
  Map,
  Skip,
  SkipWhile,
  Take,
  TakeWhile,
  Tap,
  Unique,
  Zip,
} from './adapter'

@ToString
@AsAsyncIterator
export default class Stream<T> implements AsyncProducer<T> {
  /*:: @@asyncIterator: () => $AsyncIterator<T, void, void> */
  producer: AsyncProducer<T>

  constructor(producer: AsyncProducer<T>) {
    this.producer = producer
  }

  chain<U>(source: AsyncSource<U>): Stream<T | U> {
    const adapter = new Chain(this.producer, source)
    return new Stream(adapter)
  }

  concat<U>(source: AsyncSource<U>): Stream<T | U> {
    const adapter = new Concat(this.producer, source)
    return new Stream(adapter)
  }

  async collect<C: FromIterator<T>>(Target?: Class<C>): Promise<C> {
    const items = await this.reduce((acc, next) => {
      acc.push(next)
      return acc
    }, [])

    if (Target === Array || Target == null) {
      return items
    }

    return new Target(items)
  }

  count(): Promise<number> {
    return this.reduce(acc => acc + 1, 0)
  }

  cycle(): Stream<T> {
    const adapter = new Cycle(this.producer)
    return new Stream(adapter)
  }

  drop(): void {
    this.producer.drop()
  }

  enumerate(): Stream<[number, T]> {
    const adapter = new Enumerate(this.producer)
    return new Stream(adapter)
  }

  async every(fn: T => Promise<boolean> | boolean): Promise<boolean> {
    const item = await this.find(async i => !await fn(i))
    return item === undefined
  }

  filterMap<U>(fn: T => Promise<?U> | ?U): Stream<U> {
    const adapter = new FilterMap(this.producer, fn)
    return new Stream(adapter)
  }

  filter(fn: T => Promise<boolean> | boolean): Stream<T> {
    const adapter = new Filter(this.producer, fn)
    return new Stream(adapter)
  }

  find(fn: T => Promise<boolean> | boolean): Promise<?T> {
    return this.filter(fn).first()
  }

  first(): Promise<?T> {
    return this.take(1).reduce((_, next) => next)
  }

  flatMap<U>(fn: T => AsyncSource<U>): Stream<U> {
    const adapter = new FlatMap(this.producer, fn)
    return new Stream(adapter)
  }

  flatten(): Stream<*> {
    return this.flatMap(identity)
  }

  async forEach(fn: T => Promise<void> | void): Promise<void> {
    await this.reduce((_, item) => fn(item))
  }

  join(sep?: string = ','): Promise<string> {
    return this.reduce((prev, next) => {
      if (prev.length > 0) {
        return prev + sep + String(next)
      }

      return String(next)
    }, '')
  }

  last(): Promise<?T> {
    return this.reduce((_, next) => next)
  }

  map<U>(fn: T => Promise<U> | U): Stream<U> {
    const adapter = new Map(this.producer, fn)
    return new Stream(adapter)
  }

  async next(): AsyncIteratorResult<T, void> {
    const next = await this.producer.next().catch(e => {
      this.drop()
      throw e
    })

    if (next.done) {
      this.drop()
    }

    return next
  }

  async nth(idx: number): Promise<?T> {
    const [lastIdx, value] = await this.enumerate()
      .take(idx + 1)
      .last()
      .then(item => item || [])

    if (idx === lastIdx) {
      return value
    }

    return undefined
  }

  pipe(dest: AsyncDest): Sink<*> {
    return sink.from(dest).push(this)
  }

  async product(): Promise<number> {
    const [lhs, rhs] = await Promise.all([
      this.next().then(({ value }) => +value || 0),
      this.next().then(({ value }) => +value || 0),
    ])

    return this.reduce((acc, item) => acc * +item, lhs * rhs)
  }

  reduce<U>(fn: (U, T) => Promise<U> | U, acc: Promise<U> | U): Promise<U> {
    return reduceAsync(fn, acc, this)
  }

  skip(amount: number): Stream<T> {
    const adapter = new Skip(this.producer, amount)
    return new Stream(adapter)
  }

  skipWhile(fn: T => Promise<boolean> | boolean): Stream<T> {
    const adapter = new SkipWhile(this.producer, fn)
    return new Stream(adapter)
  }

  async some(fn: T => Promise<boolean> | boolean): Promise<boolean> {
    const item = await this.find(fn)
    return item !== undefined
  }

  sum(): Promise<number> {
    return this.reduce((acc, item) => acc + +item, 0)
  }

  take(amount: number): Stream<T> {
    const adapter = new Take(this.producer, amount)
    return new Stream(adapter)
  }

  takeWhile(fn: T => Promise<boolean> | boolean): Stream<T> {
    const adapter = new TakeWhile(this.producer, fn)
    return new Stream(adapter)
  }

  tap(fn: T => Promise<void> | void): Stream<T> {
    const adapter = new Tap(this.producer, fn)
    return new Stream(adapter)
  }

  unique<U>(fn?: T => Promise<U> | Promise<T> | U | T = identity): Stream<T> {
    const adapter = new Unique(this.producer, fn)
    return new Stream(adapter)
  }

  zip(source?: AsyncSource<*> = this.producer): Stream<[T, *]> {
    const adapter = new Zip(this.producer, source)
    return new Stream(adapter)
  }
}
