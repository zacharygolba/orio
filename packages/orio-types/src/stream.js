// @flow

import type { Readable, Writable } from 'stream'

import type { Drop } from './drop'
import type { AsyncIterator, AsyncIteratorResult, Source } from './iterator'

export type AsyncDest = WritableStream | Writable
export type AsyncSource<T> =
  | Promise<Source<T>>
  | AsyncIterable<T>
  | ReadableStream
  | Readable
  | Source<T>

export interface AsyncConsumer<-T> extends Drop {
  write(chunk: T): Promise<void>,
}

export interface AsyncProducer<+T>
  extends AsyncIterable<T>, AsyncIterator<T>, Drop {}

export interface ReadableSource<+T> {
  cancel(reason?: string): void,
  read(): AsyncIteratorResult<T, void>,
}

export interface WritableDest<-T> {
  abort(reason?: string): void,
  write(chunk: T): Promise<void>,
}
