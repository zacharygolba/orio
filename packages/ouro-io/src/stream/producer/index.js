// @flow

import { intoAsyncIterator, intoIterator } from 'ouro-utils'
import type { AsyncProducer, AsyncSource } from 'ouro-types'

import * as nodeStream from '../../shims/node'
import * as webStream from '../../shims/web'
import Stream from '../stream'

import Chars from './chars'
import Empty from './empty'
import Indexed from './indexed'
import Once from './once'
import Reader from './reader'
import Task from './task'
import Unbound from './unbound'

export { Chars, Empty, Indexed, Once, Reader, Task, Unbound }

export function createProducer<T>(source: AsyncSource<T>): AsyncProducer<T> {
  if (source == null) {
    return new Empty()
  }

  if (source instanceof Promise) {
    return new Task(source)
  }

  if (source instanceof Stream) {
    return source.producer
  }

  if (webStream.isReadable(source)) {
    return new Reader(webStream.getReader(source))
  }

  if (nodeStream.isReadable(source)) {
    return new Reader(nodeStream.wrapReadable(source))
  }

  {
    const iterator = intoAsyncIterator(source)

    if (iterator != null) {
      return new Unbound(iterator)
    }
  }

  if (Array.isArray(source)) {
    return new Indexed(source)
  }

  if (typeof source === 'string') {
    return new Chars(source)
  }

  {
    const iterator = intoIterator(source)

    if (iterator != null) {
      return new Unbound(iterator)
    }
  }

  return new Once((source: any))
}
