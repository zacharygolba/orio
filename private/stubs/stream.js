// @flow

export class Duplex {}
export class PassThrough {}
export class Readable {}
export class Transform {}
export class Writable {}
