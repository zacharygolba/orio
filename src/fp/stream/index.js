// @flow

import curry from '../../utils/curry'

import * as adapters from './adapters'
import * as utils from './utils'

export type { Filter, Map } from './adapters'
export { enumerate } from './adapters'
export const chain = curry(adapters.chain)
export const filterMap = curry(adapters.filterMap)
export const filter = curry(adapters.filter)
export const flatMap = curry(adapters.flatMap)
export const map = curry(adapters.map)
export const skipWhile = curry(adapters.skipWhile)
export const skip = curry(adapters.skip)
export const takeWhile = curry(adapters.takeWhile)
export const take = curry(adapters.take)
export const tap = curry(adapters.tap)
export const zip = curry(adapters.zip)

export * from './from'
export { default as range } from './range'
export { default as repeat } from './repeat'

export type { ForEach, Reduce } from './utils'
export { count, first, last, reduce, sum, product } from './utils'
export const every = curry(utils.every)
export const find = curry(utils.find)
export const forEach = curry(utils.forEach)
export const join = curry(utils.join)
export const nth = curry(utils.nth)
export const some = curry(utils.some)
