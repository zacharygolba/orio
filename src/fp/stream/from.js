// @flow

import { Readable } from 'stream'
import type { EventEmitter } from 'events'

import noop from '../../utils/noop'

import { map } from './adapters'

const once = (source, event) => new Promise(resolve => {
  source.once(event, resolve)
})

export function from<T>(source: T): AsyncIterator<T> {
  return map(noop, source)
}

export async function* fromEvent(source: EventEmitter, event: string): AsyncIterator<*> {
  for (;;) {
    // eslint-disable-next-line no-await-in-loop
    yield await once(source, event)
  }
}

export async function* fromReadable(source: Readable): AsyncIterator<Buffer | string> {
  let done = false
  const end = () => { done = true }

  source.once('end', end)
  source.once('error', end)
  source.once('close', end)

  while (!done) {
    // eslint-disable-next-line no-await-in-loop
    yield await once(source, 'data')
    console.log(done)
  }
}
