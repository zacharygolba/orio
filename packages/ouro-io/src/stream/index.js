// @flow

import type { AsyncSource } from 'ouro-types'

import Stream from './stream'
import { createProducer, Empty } from './producer'

export type { Stream }

export function empty(): Stream<any> {
  const producer = new Empty()
  return new Stream(producer)
}

export function from<T>(source?: AsyncSource<T>): Stream<T> {
  const producer = createProducer(source)
  return new Stream(producer)
}
