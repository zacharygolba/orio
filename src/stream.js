import * as stream from './fp/stream'

class Stream {
  constructor(source) {
    this.source = source
  }

  [Symbol.iterator]() {
    return this
  }

  chain(target) {
    this.source = stream.chain(target)(this.source)
    return this
  }

  count() {
    return stream.count(this)
  }

  enumerate() {
    this.source = stream.enumerate(this.source)
    return this
  }

  every(fn) {
    return stream.every(fn)(this)
  }

  filter(fn) {
    this.source = stream.filter(fn)(this.source)
    return this
  }

  filterMap(fn) {
    this.source = stream.filterMap(fn)(this.source)
    return this
  }

  find(fn) {
    return stream.find(fn)(this)
  }

  first() {
    return stream.first(this)
  }

  flatMap(fn) {
    this.source = stream.flatMap(fn)(this.source)
    return this
  }

  forEach(fn) {
    return stream.forEach(fn)(this)
  }

  join(sep) {
    return stream.join(sep)(this)
  }

  last() {
    return stream.last(this)
  }

  map(fn) {
    this.source = stream.map(fn)(this.source)
    return this
  }

  next() {
    return this.source.next()
  }

  nth(index) {
    return stream.nth(index)(this)
  }

  product() {
    return stream.product(this)
  }

  reduce(fn, init) {
    return stream.reduce(fn, init, this)
  }

  skip(amount) {
    this.source = stream.skip(amount)(this.source)
    return this
  }

  skipWhile(fn) {
    this.source = stream.skipWhile(fn)(this.source)
    return this
  }

  some(fn) {
    return stream.some(fn)(this)
  }

  sum() {
    return stream.sum(this)
  }

  take(amount) {
    this.source = stream.take(amount)(this.source)
    return this
  }

  takeWhile(fn) {
    this.source = stream.takeWhile(fn)(this.source)
    return this
  }

  tap(fn) {
    this.source = stream.tap(fn)(this.source)
    return this
  }

  zip(target) {
    this.source = stream.zip(target)(this.source)
    return this
  }
}

export function from(source) {
  return new Stream(stream.from(source))
}

export function fromEvent(source, event) {
  return new Stream(stream.fromEvent(source, event))
}

export function fromReadable(source) {
  return new Stream(stream.fromReadable(source))
}

export function range(start, end) {
  return new Stream(stream.range(start, end))
}

export function repeat(value) {
  return new Stream(stream.repeat(value))
}
