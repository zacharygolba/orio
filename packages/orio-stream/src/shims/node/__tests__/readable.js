// @flow

import { Readable } from 'stream'

import ReadableWrapper from '../readable'

let data: Buffer
let wrapper: ReadableWrapper<Buffer>

beforeEach(() => {
  let offset = 0

  data = Buffer.from('Hello World')
  wrapper = new ReadableWrapper(
    new Readable({
      read(size) {
        if (offset >= data.length) {
          this.push(null)
        } else {
          this.push(data.slice(offset, offset + size))
          offset += size
        }
      },
    }),
  )

  jest.spyOn(wrapper, 'cancel')
  jest.spyOn(wrapper.source, 'unpipe')
  jest.spyOn(wrapper.pipe, 'removeAllListeners')
})

test('#constructor()', () => {
  expect(wrapper).toMatchSnapshot()
})

test('#cancel()', async () => {
  expect(() => wrapper.cancel()).not.toThrow()
  expect(wrapper.source.unpipe).toHaveBeenCalled()
  expect(wrapper.pipe.removeAllListeners).toHaveBeenCalled()

  {
    const { done, value } = await wrapper.read()
    expect(done).toBe(true)
    expect(value).toBeUndefined()
  }
})

describe('#read()', () => {
  test('success', async () => {
    {
      const { done, value } = await wrapper.read()
      expect(done).toBe(false)
      expect(value).toEqual(data)
    }

    {
      const { done, value } = await wrapper.read()
      expect(done).toBe(true)
      expect(value).toBeUndefined()
    }
  })

  test('failure', async () => {
    const error = new Error('test')

    wrapper.pipe.emit('error', error)
    await wrapper.read().catch(e => {
      expect(e).toBe(error)
      expect(wrapper.cancel).toHaveBeenCalled()
    })
  })

  test('upstream failure', async () => {
    const error = new Error('test')

    wrapper.source.emit('error', error)
    await wrapper.read().catch(e => {
      expect(e).toBe(error)
      expect(wrapper.cancel).toHaveBeenCalled()
    })
  })
})
