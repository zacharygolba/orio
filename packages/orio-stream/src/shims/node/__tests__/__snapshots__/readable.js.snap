// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#constructor() 1`] = `
ReadableWrapper {
  "cancel": [Function],
  "pipe": ReadablePipe {
    "_events": Object {
      "close": [Function],
      "drain": [Function],
      "end": [Function],
      "error": Array [
        [Function],
        [Function],
        [Function],
      ],
      "finish": [Function],
      "prefinish": [Function],
      "unpipe": [Function],
    },
    "_eventsCount": 7,
    "_maxListeners": undefined,
    "_readableState": ReadableState {
      "awaitDrain": 0,
      "buffer": BufferList {
        "head": null,
        "length": 0,
        "tail": null,
      },
      "decoder": null,
      "defaultEncoding": "utf8",
      "destroyed": false,
      "emittedReadable": false,
      "encoding": null,
      "endEmitted": false,
      "ended": false,
      "flowing": false,
      "highWaterMark": 16384,
      "length": 0,
      "needReadable": true,
      "objectMode": false,
      "pipes": null,
      "pipesCount": 0,
      "readableListening": false,
      "reading": false,
      "readingMore": false,
      "resumeScheduled": false,
      "sync": false,
    },
    "_transformState": Object {
      "afterTransform": [Function],
      "needTransform": false,
      "transforming": false,
      "writecb": null,
      "writechunk": null,
      "writeencoding": null,
    },
    "_writableState": WritableState {
      "bufferProcessing": false,
      "bufferedRequest": null,
      "bufferedRequestCount": 0,
      "corked": 0,
      "corkedRequestsFree": Object {
        "entry": null,
        "finish": [Function],
        "next": null,
      },
      "decodeStrings": true,
      "defaultEncoding": "utf8",
      "destroyed": false,
      "ended": false,
      "ending": false,
      "errorEmitted": false,
      "finalCalled": false,
      "finished": false,
      "highWaterMark": 16384,
      "lastBufferedRequest": null,
      "length": 0,
      "needDrain": false,
      "objectMode": false,
      "onwrite": [Function],
      "pendingcb": 0,
      "prefinished": false,
      "sync": true,
      "writecb": null,
      "writelen": 0,
      "writing": false,
    },
    "allowHalfOpen": true,
    "domain": null,
    "done": false,
    "error": undefined,
    "readable": true,
    "removeAllListeners": [Function],
    "writable": true,
  },
  "source": Readable {
    "_events": Object {
      "data": [Function],
      "end": [Function],
      "error": [Function],
    },
    "_eventsCount": 3,
    "_maxListeners": undefined,
    "_read": [Function],
    "_readableState": ReadableState {
      "awaitDrain": 0,
      "buffer": BufferList {
        "head": null,
        "length": 0,
        "tail": null,
      },
      "decoder": null,
      "defaultEncoding": "utf8",
      "destroyed": false,
      "emittedReadable": false,
      "encoding": null,
      "endEmitted": false,
      "ended": false,
      "flowing": true,
      "highWaterMark": 16384,
      "length": 0,
      "needReadable": false,
      "objectMode": false,
      "pipes": ReadablePipe {
        "_events": Object {
          "close": [Function],
          "drain": [Function],
          "end": [Function],
          "error": Array [
            [Function],
            [Function],
            [Function],
          ],
          "finish": [Function],
          "prefinish": [Function],
          "unpipe": [Function],
        },
        "_eventsCount": 7,
        "_maxListeners": undefined,
        "_readableState": ReadableState {
          "awaitDrain": 0,
          "buffer": BufferList {
            "head": null,
            "length": 0,
            "tail": null,
          },
          "decoder": null,
          "defaultEncoding": "utf8",
          "destroyed": false,
          "emittedReadable": false,
          "encoding": null,
          "endEmitted": false,
          "ended": false,
          "flowing": false,
          "highWaterMark": 16384,
          "length": 0,
          "needReadable": true,
          "objectMode": false,
          "pipes": null,
          "pipesCount": 0,
          "readableListening": false,
          "reading": false,
          "readingMore": false,
          "resumeScheduled": false,
          "sync": false,
        },
        "_transformState": Object {
          "afterTransform": [Function],
          "needTransform": false,
          "transforming": false,
          "writecb": null,
          "writechunk": null,
          "writeencoding": null,
        },
        "_writableState": WritableState {
          "bufferProcessing": false,
          "bufferedRequest": null,
          "bufferedRequestCount": 0,
          "corked": 0,
          "corkedRequestsFree": Object {
            "entry": null,
            "finish": [Function],
            "next": null,
          },
          "decodeStrings": true,
          "defaultEncoding": "utf8",
          "destroyed": false,
          "ended": false,
          "ending": false,
          "errorEmitted": false,
          "finalCalled": false,
          "finished": false,
          "highWaterMark": 16384,
          "lastBufferedRequest": null,
          "length": 0,
          "needDrain": false,
          "objectMode": false,
          "onwrite": [Function],
          "pendingcb": 0,
          "prefinished": false,
          "sync": true,
          "writecb": null,
          "writelen": 0,
          "writing": false,
        },
        "allowHalfOpen": true,
        "domain": null,
        "done": false,
        "error": undefined,
        "readable": true,
        "removeAllListeners": [Function],
        "writable": true,
      },
      "pipesCount": 1,
      "readableListening": false,
      "reading": false,
      "readingMore": false,
      "resumeScheduled": true,
      "sync": true,
    },
    "domain": null,
    "readable": true,
    "unpipe": [Function],
  },
}
`;
